xarray/core/dataset.py

```diff
@@ -5052,8 +5052,7 @@ def _assert_all_in_dataset(
             bad_names -= self.virtual_variables
         if bad_names:
             raise ValueError(
-                "One or more of the specified variables "
-                "cannot be found in this dataset"
+                f"These variables cannot be found in this dataset: {bad_names}"
```

Comment:
I'm not sure if this is possible, but the use of `set` means that there is no order within `bad_names`: different python session will give you a different order and thus a different error message (which is bad for `doctest`s, in particular).

I'd usually sort `bad_names`, but since we allow arbitrary hashables instead of just strings we can't do that. Instead, how about ordering by `_variables`? For example, we could do:
```python
bad_names_ = [name for name in self._variables if name in bad_names]
```
to get the same order as the dataset's variables.

I'm not sure what to do with virtual variables, but if I remember correctly, those have been removed in the index refactor, anyways? Can you confirm, @benbovy?

Reply:
> I'm not sure what to do with virtual variables, but if I remember correctly, those have been removed in the index refactor, anyways? Can you confirm, @benbovy?

Not all of them, e.g., `time.year` / `time.month` / etc. are still there.

Reply:
actually, never mind: anything in `bad_names` is by definition not in `self.variables` or `self.virtual_variables`, so we can just keep the order in `names`:
```python
bad_names_ = [name for name in names if name in bad_names]
```

Reply:
Good point - now resolved.